### PR TITLE
Copied colleague's unique pointer implementation to prepare for review

### PR DESCRIPTION
**Pull request by kyorri**
resolved the memory deallocation problem in the main function and added the rule of five for the room class

hw6
---
added the UniquePointer implementation using a Singleton design pattern

### CODE REVIEW

File: smart-home-application/include/Room.h [hunk 1]
```diff
@@ -3,24 +3,34 @@
 
 #include "Device.h"
 
+
```
Comment (CristianIonutIonescu) [inline smart-home-application/include/Room.h]: unnecessary extra new line

File: smart-home-application/src/Reader.cpp [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#include "Reader.h"
+
+namespace smart_home_app {
+
+} // smart_home_app
```
Comment (CristianIonutIonescu) [inline smart-home-application/src/Reader.cpp]: this file is not needed

File: smart-home-application/src/Writer.cpp [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#include "Writer.h"
+
+namespace smart_home_app {
+    
+} // namespace smart_home_app
```
Comment (CristianIonutIonescu) [inline smart-home-application/src/Writer.cpp]: same as reader, not needed

File: smart-home-application/include/UniquePointer.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#ifndef SHA_UNIQUEPOINTER_H_
+#define SHA_UNIQUEPOINTER_H_
+
+#include <memory>
+
+namespace up {
+
+    template<typename T>
+    class UniquePointer {
```
Comment (CristianIonutIonescu) [inline smart-home-application/include/UniquePointer.h]: unique_ptr class has the ability to use a custom deleter as well, please double check the definition of [std::unique_ptr](https://en.cppreference.com/w/cpp/memory/unique_ptr)

File: smart-home-application/src/Room.cpp [hunk 5]
```diff
@@ -4,6 +4,55 @@
...
+        return room_name_;
+    };
+
+    void Room::AddDevice(Device* new_device) {
+        devices_.push_back(new_device);
+    };
+
+    void Room::RemoveDevice(Device* rem_device) {
+        auto it = std::find(devices_.begin(), devices_.end(), rem_device);
+        devices_.erase(it);
```
Comment (CristianIonutIonescu) [inline smart-home-application/src/Room.cpp]: you have a bug here but we'll discuss in the next presentation about this

File: smart-home-application/src/Room.cpp [hunk 6]
```diff
@@ -4,6 +4,55 @@
...
+        for (Device* d : devices_) {
+            delete d;
+        }
+        devices_.clear();
+    };
+
+    Room::Room(const Room& other) : room_name_(other.room_name_), devices_(other.devices_) {};
+
+    Room& Room::operator=(const Room& other) {
+        Room tmp(other);
```
Comment (CristianIonutIonescu) [inline smart-home-application/src/Room.cpp]: incorrect implementation of copy assignment operator, need to check first if other is different than *this

File: smart-home-application/src/Room.cpp [hunk 7]
```diff
@@ -4,6 +4,55 @@
...
+        std::swap(room_name_, tmp.room_name_);
+        std::swap(devices_, tmp.devices_);
+        return *this;
+    };
+    
+    Room::Room(Room&& other) : room_name_(std::move(other.room_name_)), devices_(std::move(other.devices_)) {};
+
+    Room& Room::operator=(Room&& other) {
+        room_name_ = std::move(other.room_name_);
```
Comment (CristianIonutIonescu) [inline smart-home-application/src/Room.cpp]: incorrect implementation of move assignment operator, need to check first if other is different than *this

File: smart-home-application/src/TemperatureSensor.cpp [hunk 8]
```diff
@@ -27,6 +31,8 @@ namespace smart_home_app {
     void TemperatureSensor::SetScaleTo(char initial_scale, char final_scale) {
         double celsius_value;
         double final_value;
+        celsius_value = 0.0;
+        final_value = 0.0;
```
Comment (CristianIonutIonescu) [inline smart-home-application/src/TemperatureSensor.cpp]: why not:
````
double celsius_value = 0.0
double final_value = 0.0
````

?

File: smart-home-application/src/UniquePointer.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
+#include "UniquePointer.h"
```
Comment (CristianIonutIonescu) [inline smart-home-application/src/UniquePointer.cpp]: this file needs to be removed, cannot declare template classes methods in header and define them in source files

**Review by CristianIonutIonescu — CHANGES_REQUESTED** [review]
Please fix highlighted issues

**Comment by kyorri** [comment]
- Added my custom UniquePointer implementation using the [cpp reference ](https://en.cppreference.com/w/cpp/memory/unique_ptr) and a simple custom Deleter.
- Resolved the other findings.

**Comment by kyorri** [comment]
> Added also the exception and error handling to the code
> Updated the code to use smart pointers (shared_ptr only at the moment)

**Review by CristianIonutIonescu — APPROVED** [review]

**Review by CristianIonutIonescu — CHANGES_REQUESTED** [review]
scratch my previous approval.
Got multiple emails and thought that you just copied Antonia's unique_ptr and moved some files around, seems it's not the case, you're changing on top of homework 5 PR.
Please do the following:
- rollback your previous commits here to have just what homework 5 needed to have, I'll double check everything is ok and I'll aprove it. Afterwards you merge it.
- Copy Antonia's unique_ptr implementation in a new branch and create a PR with it. I'll approve it and afterwards you merge it
- Create a new branch, where you add only the changes for your current homework, no other commits, please. It is very hard to follow your PRs

File: smart-home/src/Room.cpp [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    void Room::RemoveDevice(Device* rem_device) {
+        auto it = std::find(devices_.begin(), devices_.end(), rem_device);
+        devices_.erase(it);
+    };
+
+    bool Room::operator==(const Room& other) {
+        return ((this->devices_ == other.devices_) && (this->room_name_ == other.room_name_));
+    }
+
+    std::vector<Device*> Room::GetDevices() {
```
Comment (enachebogdan) [inline smart-home/src/Room.cpp]: you should try returning a const & to your vector instead of a copy. some people prefer to return const references for any non-primitive type instead of copying.

File: smart-home/src/TemperatureSensor.cpp [hunk 11]
```diff
@@ -4,6 +4,10 @@
 
 namespace smart_home_app {
 
+    std::string TemperatureSensor::GetType() {
+        return "temp";
+    };
+
     std::string TemperatureSensor::GetInfo() {
         std::string message = std::string();
         message += "This Temperature Sensor is reading " + std::to_string(GetReading()) + " degrees " + GetScale() + "!\n";
```
Comment (enachebogdan) [inline smart-home/src/TemperatureSensor.cpp]: why don't you do directly "std::string message = "This Temperature.."?

File: smart-home/src/XMLReader.cpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                        ts->SetScale(*scale.c_str());
+                    }
+                    else if (sensor_type == "hum") {
+                        // Write derived attributes in case of extension of hum
+                    }
+                    else if (sensor_type == "dist") {
+                        std::string unit = sensor_node.attribute("unit").value();
+                        DistanceSensor* ds = dynamic_cast<DistanceSensor*>(s);
+                        ds->SetUnits(unit);
+                    }
```
Comment (enachebogdan) [inline smart-home/src/XMLReader.cpp]: if the sensor_type string is none of the above, do you want to add a nullptr to your vector?
it's advised to add a final "else" that will handle this case. Similar for the Device parsing.

File: smart-home/src/XMLWriter.cpp [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    for (Room r : rooms_) {
+        pugi::xml_node room_root = root.append_child("Room");
+        room_root.append_attribute("name").set_value(r.GetRoomName().c_str());
+
+        for (Device* d : r.GetDevices()) {
+            pugi::xml_node device_root = room_root.append_child("Device");
+            device_root.append_attribute("type").set_value(d->GetType().c_str());
+            device_root.append_attribute("status").set_value(d->GetStatus().c_str());
+
+            std::string device_type = d->GetType();
```
Comment (enachebogdan) [inline smart-home/src/XMLWriter.cpp]: No need to change anything for this homework, just something to consider for the future. Your code is not very SOLID friendly, a xml writer class should not care about your data types. should be something like:
for(device*d : getDevices) { device_root.append_attribute(d->GetData())} or something similar. If you have 50 types of devices and 100 of Sensors, you would modify your code for each of them.

File: smart-home/src/XMLWriter.cpp [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+void XMLWriter::Write(std::string filename) {
+    pugi::xml_document doc;
+    pugi::xml_node root = doc.append_child("Root");
+
+    for (Room r : rooms_) {
+        pugi::xml_node room_root = root.append_child("Room");
+        room_root.append_attribute("name").set_value(r.GetRoomName().c_str());
+
+        for (Device* d : r.GetDevices()) {
+            pugi::xml_node device_root = room_root.append_child("Device");
```
Comment (enachebogdan) [inline smart-home/src/XMLWriter.cpp]: It's usually a good idea to check if a pointer is nullptr before accessing it (even if for your case this didn't occur). similar for Sensor.

File: smart-home/src/XMLReader.cpp [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+                std::string device_type = device_node.attribute("type").value();
+                std::string device_status = device_node.attribute("status").value();
+
+                if (device_type == "acunit") {
+                    // Write derived info in case of extension for acunit
+                }
+                else if (device_type == "doorbell") {
+                    std::string visitors = device_node.attribute("visitors").value();
+                    dynamic_cast<DoorbellDevice*>(d)->SetVisitors(std::stoi(visitors));
```
Comment (enachebogdan) [inline smart-home/src/XMLReader.cpp]: I am not sure what you are trying to do with the dynamic cast here. Is your code working? have you checked if your vector of devices has any valid elements?
Your allocation/initialization should be very clear. And please use smart pointers(could have used the unique pointer of your colleague)

**Review by enachebogdan — COMMENTED** [review]
Homework looks good, I have added few questions and notes.

File: smart-home/src/XMLReader.cpp (hunk 15 above)
Comment (kyorri) [reply smart-home/src/XMLReader.cpp]: Some devices and sensors have uncommon members in their functions (the number of visitors, in this case), I did dynamically cast the read device as I would access the setter method of the DoorbellDevice to modify the value of the visitors.

File: smart-home/src/XMLWriter.cpp (hunk 14 above)
Comment (kyorri) [reply smart-home/src/XMLWriter.cpp]: I understand, now I am just throwing an exception in case of an error if the pointer is nullptr, so that I would not go about trying to write a Device@nullptr.

File: smart-home/src/TemperatureSensor.cpp (hunk 11 above)
Comment (kyorri) [reply smart-home/src/TemperatureSensor.cpp]: In the earlier versions of the sensors/devices of the code, I would take into account the type of sensors/devices as a string, thinking I would need somewhere to specify the type of sensor/device before some other operation, such as the writing to XML files.

File: smart-home/src/XMLWriter.cpp (hunk 13 above)
Comment (kyorri) [reply smart-home/src/XMLWriter.cpp]: I understand, I will try to be more careful next time when it comes to the overall structure of my classes.

File: smart-home/src/Room.cpp (hunk 10 above)
Comment (kyorri) [reply smart-home/src/Room.cpp]: I understand that I will pass some large resources around, I will make sure to use this construct more frequently.

File: smart-home/src/XMLReader.cpp (hunk 12 above)
Comment (kyorri) [reply smart-home/src/XMLReader.cpp]: Now, when I am reading Devices/Sensors, I will take into account the DeviceType/SensorType which could be Undefined, in the case of a mistaken read.

**Review by CristianIonutIonescu — APPROVED** [review]
